app/code/community/Twispay/Tpay/Model/PaymentMethod.php

```diff
@@ -1,64 +1,222 @@
+   *  option is set to "Authorize and Capture" and will automatically
+   *  authorize and capture all the received orders.
+   * 
+   * Background operations:
+   *   - order with status PROCESSING is created;
+   *   - transaction is registered into magento with the transaction id;
+   *   - order comment is added automatically for authorization;
+   *   - invoice is created with status PAID;
+   */
+  public function capture(Varien_Object $payment, $amount) {
+    Mage::Log(Mage::helper('tpay')->__(' Authorize and capture payment'), Zend_Log::DEBUG, $this->logFileName);
+    return $this;
+  }
+
+
+  /**
+   * Function that is called when a refund is done.
+   */
+  public function refund(Varien_Object $payment, $amount) {
+    Mage::Log(__FUNCTION__ . ': amount=' . print_r($amount, true), Zend_Log::DEBUG, $this->logFileName);
+
+    /* Extract the transaction and transaction data. */
+    $transactionId = $this->_getParentTransactionId($payment);
+    $transaction = $payment->getTransaction($transactionId);
+    $transactionData = $transaction->getData()['additional_information']['raw_details_info'];
+
+    /* Get the config values. */
+    $apiKey = Mage::helper('tpay')->getApiKey();
+    Mage::Log(__FUNCTION__ . ': apiKey=' . print_r($apiKey, true), Zend_Log::DEBUG, $this->logFileName);
+    $url = Mage::helper('tpay')->getApiUrl();
+    Mage::Log(__FUNCTION__ . ': url=' . print_r($url, true), Zend_Log::DEBUG, $this->logFileName);
+
+    if ('' == $apiKey) {
+      Mage::getSingleton('core/session')->addError(Mage::helper('tpay')->__(' Refund failed: Incomplete or missing configuration.'));
+      $this->_redirect('adminhtml/sales_order/view', $transactionData['orderId']);
     }
 
-    // Construct the redirect URL
-    public function getOrderPlaceRedirectUrl()
-    {   
-        $redirect_url = Mage::getUrl('tpay/payment/redirect');
-        Mage::Log("Step 2 Process: Getting the redirect URL: $redirect_url", Zend_Log::DEBUG, $this->LOG_FILE_NAME);
-        return $redirect_url;      
+    /* Create the URL. */
+    $url = $url . '/transaction/' . $transactionId;
+
+    /* Create the DELETE data arguments. */
+    $postData = 'amount=' . $amount . '&' . 'message=' . 'Refund for order ' . $transactionData['orderId'];
+
+    /* Make the server request. */
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL, $url);
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "DELETE");
+    curl_setopt($ch, CURLOPT_HTTPHEADER, ['accept: application/json', 'Authorization: ' . $apiKey]);
+    curl_setopt($ch, CURLOPT_POST, count($postData));
+    curl_setopt($ch, CURLOPT_POSTFIELDS, $postData);
+
+    /* Send the request. */
+    $response = curl_exec($ch);
+    curl_close($ch);
+    /* Decode the response. */
+    $response = json_decode($response);
+
+    /* Check if the response code is 200 and message is 'Success'. */
+    if ((200 == $response->code) && ('Success' == $response->message)) {
```

Comment:
We should also check if the response was decoded successfully with `json_decode`.

Reply:
Added the check.